opentracing-impl-java8/src/main/java/io/opentracing/propagation/AbstractTracer.java

```diff
@@ -0,0 +1,175 @@
+                    return extractors.get(c);
+                }
+            }
+            throw new AssertionError("no registered extractor for " + carrierType.getName());
+        }
+
+        public <T> Injector<T> register(Class<T> carrierType, Injector<T> injector) {
+            return injectors.putIfAbsent(carrierType, injector);
+        }
+
+        public <T> Extractor<T> register(Class<T> carrierType, Extractor<T> extractor) {
+            return extractors.putIfAbsent(carrierType, extractor);
+        }
+    }
+
+    static class SpanBuilderImpl implements SpanBuilder {
+
+        SpanBuilderImpl() {}
+
+        final Map<String,String> traceState = new HashMap<>();
+        final Map<String,String> traceAttributes = new HashMap<>();
+
+        private String operationName = null;
+        private Span parent = null;
+        private Instant start = Instant.now();
+        private final Map<String,String> stringTags = new HashMap<>();
+        private final Map<String,Boolean> booleanTags = new HashMap<>();
+        private final Map<String,Number> numberTags = new HashMap<>();
+
+        @Override
+        public Tracer.SpanBuilder withOperationName(String operationName) {
+            this.operationName = operationName;
+            return this;
+        }
+
+        @Override
+        public Tracer.SpanBuilder withParent(Span parent) {
+            this.parent = parent;
+            return this;
+        }
+
+        @Override
+        public Tracer.SpanBuilder withTag(String key, String value) {
+            stringTags.put(key, value);
+            return this;
+        }
+
+        @Override
+        public Tracer.SpanBuilder withTag(String key, boolean value) {
+            booleanTags.put(key, value);
+            return this;
+        }
+
+        @Override
+        public Tracer.SpanBuilder withTag(String key, Number value) {
+            numberTags.put(key, value);
+            return this;
+        }
+
+        @Override
+        public Tracer.SpanBuilder withStartTimestamp(long microseconds) {
+            long epochSeconds = TimeUnit.MICROSECONDS.toSeconds(microseconds);
+            long nanos = TimeUnit.MICROSECONDS.toNanos(microseconds) - TimeUnit.SECONDS.toNanos(epochSeconds);
+            this.start = Instant.ofEpochSecond(epochSeconds, nanos);
+            return this;
+        }
+
+        @Override
+        public Span start() {
+            Span span = new SpanImpl(operationName, Optional.ofNullable(parent), start);
+            stringTags.entrySet().stream().forEach((entry) -> { span.setTag(entry.getKey(), entry.getValue()); });
+            booleanTags.entrySet().stream().forEach((entry) -> { span.setTag(entry.getKey(), entry.getValue()); });
+            numberTags.entrySet().stream().forEach((entry) -> { span.setTag(entry.getKey(), entry.getValue()); });
+            traceState.entrySet().stream().forEach((entry) -> { span.setBaggageItem(entry.getKey(), entry.getValue()); });
```

Comment:
why not just pass already constructed maps for tags and baggage to span constructor?